runtime/drivers/druid/druidsqldriver/druid_api_sql_driver.go

```diff
@@ -40,187 +55,286 @@ type sqlConnection struct {
 	if err != nil {
-		return nil, err
+		return false, err
 	}
 
 	req.Header.Add("Content-Type", "application/json")
-	resp, err := c.client.Do(req)
+	resp, err := chc.c.client.Do(req)
 	if err != nil {
-		return nil, err
+		return false, err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusTooManyRequests:
+		return false, nil
+	case http.StatusUnauthorized, http.StatusForbidden:
+		return true, fmt.Errorf("Unauthorized request")
 	}
 
 	dec := json.NewDecoder(resp.Body)
 
 	var obj any
 	err = dec.Decode(&obj)
+	resp.Body.Close()
 	if err != nil {
-		resp.Body.Close()
-		return nil, err
+		return false, err
 	}
 	switch v := obj.(type) {
 	case map[string]any:
-		resp.Body.Close()
-		return nil, fmt.Errorf("%v", obj)
+		if v["errorCode"] != "invalidInput" {
+			return false, fmt.Errorf("%v", obj)
+		}
+		return true, nil
 	case []any:
-		columns := toStringArray(v)
+		return true, nil
+	default:
+		return false, fmt.Errorf("unexpected response: %v", obj)
+	}
+}
+
+func (c *sqlConnection) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	// total sum is 126 seconds (sum(2*2^x) from 0 to 5 inclusive)
+	re := retrier.NewRetrier(6, 2*time.Second, &coordinatorHTTPCheck{
+		c: c,
+	})
+	return re.RunCtx(ctx, func(ctx2 context.Context) (driver.Rows, retrier.Action, error) {
+		dr := newDruidRequest(query, args)
+		b, err := json.Marshal(dr)
+		if err != nil {
+			return nil, retrier.Fail, err
+		}
+
+		bodyReader := bytes.NewReader(b)
+
+		context.AfterFunc(ctx, func() {
+			tctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			r, err := http.NewRequestWithContext(tctx, http.MethodDelete, c.dsn+"/"+dr.Context.SQLQueryID, http.NoBody)
+			if err != nil {
+				return
+			}
+
+			resp, err := c.client.Do(r)
+			if err != nil {
+				return
+			}
+			resp.Body.Close()
+		})
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.dsn, bodyReader)
+		if err != nil {
+			return nil, retrier.Fail, err
+		}
+
+		req.Header.Add("Content-Type", "application/json")
+		resp, err := c.client.Do(req)
+		if err != nil {
+			// nolint:errorlint // there's no wrapping
+			if v, ok := err.(*url.Error); ok {
+				if tooManyRedirects.MatchString(v.Error()) {
+					return nil, retrier.Fail, v
+				}
+
+				if invalidProtocol.MatchString(v.Error()) {
+					return nil, retrier.Fail, v
+				}
+
+				if TLSCert.MatchString(v.Error()) {
+					return nil, retrier.Fail, v
+				}
+
+				// nolint:errorlint // there's no wrapping
+				if _, ok := v.Err.(x509.UnknownAuthorityError); ok {
+					return nil, retrier.Fail, v
+				}
+			}
+
+			return nil, retrier.Retry, err
+		}
+
+		switch resp.StatusCode {
+		case http.StatusTooManyRequests:
+			return nil, retrier.Retry, fmt.Errorf("Too many requests")
+		case http.StatusUnauthorized, http.StatusForbidden:
+			return nil, retrier.Fail, fmt.Errorf("Unauthorized request")
+		}
```

Comment:
I've added `defer` in the other request. Here the body should go to a handler that will parse it outside of the function so that we can support `drivers.Rows.Next()` semantics.